enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/assets/k8s/Pods.java

```diff
@@ -0,0 +1,165 @@
+import java.util.concurrent.ExecutionException;
+import org.netbeans.api.project.Project;
+import org.netbeans.api.project.ProjectInformation;
+import org.netbeans.api.project.ProjectUtils;
+import org.netbeans.modules.cloud.oracle.assets.OpenProjectsFinder;
+import org.openide.util.Exceptions;
+import org.openide.util.RequestProcessor;
+
+/**
+ *
+ * @author Jan Horvath
+ */
+public class Pods {
+    private final RequestProcessor RP = new RequestProcessor(Pods.class);
+    private final ClusterItem cluster;
+    private Set<String> names = null;
+    private final List<PodItem> pods = new ArrayList<> ();
+    private transient PropertyChangeSupport changeSupport = null;
+    private CountDownLatch latch = null;
+    
+    private Pods(ClusterItem cluster) {
+        this.cluster = cluster;
+    }
+    
+    public static Pods from(ClusterItem cluster) {
+        Pods instance = new Pods(cluster);
+        instance.init();
+        return instance;
+    }
+    
+    public List<PodItem> getItems() {
+        return Collections.unmodifiableList(pods);
+    }
+    
+    private void addPod(PodItem pod) {
+        pods.add(pod);
+        changeSupport.firePropertyChange("pods", pods.size(), pods.size() - 1);
+        
+    }
+    
+    private void removePod(PodItem pod) {
+        pods.remove(pod);
+        PortForwards.getDefault().closePortForward(pod);
+        changeSupport.firePropertyChange("pods", pods.size(), pods.size() + 1);
+    }
+    
+    private void removeAll() {
+        pods.clear();
+        changeSupport.firePropertyChange("pods", pods.size(), pods.size() + 1);
+    }
+
+    private void init() {
+        
+        try {
+            changeSupport = new PropertyChangeSupport(this);
+            cluster.addChangeListener((PropertyChangeEvent evt) -> {
+                if ("namespace".equals(evt.getPropertyName())) {
+                    watchNamespace((String) evt.getNewValue());
+                }
+            });
+            CompletableFuture<Project[]> projectsFuture = OpenProjectsFinder.getDefault().findTopLevelProjects();
+            projectsFuture.thenApply(projects -> {
+                List<String> projectNames = new ArrayList<>();
+                for (int i = 0; i < projects.length; i++) {
+                    ProjectInformation pi = ProjectUtils.getInformation(projects[i]);
+                    projectNames.add(pi.getDisplayName());
+                }
+                return projectNames;
+            }).thenAccept(projectNames -> {
+                names = new HashSet<> (projectNames);
+                
+            }).get();
+            watchNamespace(cluster.getNamespace());
+        } catch (InterruptedException | ExecutionException ex) {
+            Exceptions.printStackTrace(ex);
+        }
+    }
+    
+    private void watchNamespace(String namespace) {
+        removeAll();
+        if (latch != null) {
+            latch.countDown();
+        }
+        latch = new CountDownLatch(1);
```

Comment:
watchNamespace() should synchronize access to `latch`;  the new latch should be copied to local variable before synchronized block terminates, Watcher::onClose  should use that local copy.